### PR TITLE
allow ansible-pull to specify playbook; improve option parsing

### PR DESCRIPTION
**Pull request by copperlight**
This branch adds a "-P, --playbook" parameter to ansible-pull, which makes it easier to specify a custom playbook that should run on a pull-based node.  The new parameter is optional, and the top-level local.yml file is still the default.  This concept behind adding this parameter is intended to help with scale-out so that you can have many nodes that each run against their own custom playbook, so that you can have distinctions between web servers and database servers, for example.

Within a playbook repo, a new directory structure houses the custom playbooks, per node:

nodes/${ansible_fqdn}.yml

ansible-pull is called against it from a cronjob, running every 30m:

ansible-pull -d /var/lib/ansible -U https://${username}:${password}@github.com/${username}/${repo}.git -P nodes/${ansible_fqdn}.yml

ansible-pull will now spit out a starting date and time "ansible-pull_started:", to help with reading the log file output over time.

Example playbook updates to support this updated ansible-pull are bundled into a separate pull request.

### CODE REVIEW

**Comment by mpdehaan** [comment]
-P sounds good, in queue for review/testing a bit later

**Comment by copperlight** [comment]
Cool! :)  Testing on my Ubuntu 12.04 box went well, I will keep an eye out for further suggestions.

**Comment by sfromm** [comment]
Instead of adding a -P option, how about making ansible-pull operate a little more similar to ansible-playbook.  Make ansible-pull optionally take the name of a playbook to run.  Such as:

```
ansible-pull [options] $hostname.yml
```

This way, someone can choose to use $hostname.yml or $groupname.yml. If a playbook file isn't provided, it defaults to local.yml.  Thoughts?

**Comment by copperlight** [comment]
That's not a bad way to do that... it makes the command line implementation more concise.

mpdehaan suggested an auto mode that would do the following, in order: (1) use the provided playbook name, if it exists, (2) search for $hostname.yml in a pre-defined location, (3) fall back to local.yml at the top level of the repo.

One open question regarding auto-searching for $hostname.yml -- should it exist at the top level of the repo or be pushed down into a "nodes" directory?

**Comment by sfromm** [comment]
I would say the top-level.  If the user wants to separate these to a subdirectory, she/he can specify _ansible-pull [options] nodes/$hostname.yml_.

**Comment by copperlight** [comment]
Works for me, I'll get started on those changes.

**Comment by copperlight** [comment]
Playbook auto-detection implemented.  It's a bit wordy on the output, but I think the slight bit of extra verbosity will help people wrap their heads around what is happening when it operates.

**Comment by sfromm** [comment]
The way you derive hostname is different than how setup does it.  Not sure if that will lead to a mismatch.  For example, a host with the fqdn foo.bar.example.com may have foo or foo.bar as a legitimate hostname.

The code appears to look for a file named $hostname, not $hostname.yml.  It would also be nice if the two duplicate blocks of code could be collapsed into one.  Moving the git operation before the check for the playbook is a nice improvement.

Not sure if this is bikeshedding, but I'd rather see auto-detection based on the fqdn.  If people try to deploy across multiple domains, or sub-domains, there is a risk of name-overload because of the lack of uniqueness in hostnames. Using the fqdn avoids the potential vagaries.

**Comment by copperlight** [comment]
I agree that FQDN will be a better way to describe and search for these files.  Updated to match the method used in ansible/libary/setup.

This change does not require updating the playbook example, since it was simplified.

Code is searching specifically for $hostname.yml, line 95 builds the name to be checked by file open actions:

```
hostname = "%s.yml" % socket.getfqdn()
```

While it would be nice to collapse the playbook auto-search code blocks, I can't think of any clever way to do it.  Stripping the text output on success and failure would make it tighter, but less obvious when something goes wrong in unattended execution.

It is essentially this:

```
if (no playbook):
  does $hostname.yml exist?
  does local.yml exist?
else:
  does playbook.yml exist?
  does $hostname.yml exist?
  does local.yml exist?
```

If you can think of a better way to collapse it, I can implement it.

**Comment by sfromm** [comment]
How about (ugly pseudo python to follow):

```
def auto_detect_playbook():
   try $hostname.yml
   try localyml

def main():
   ...
   if arg:
      # blah
   else:
      auto_detect_playbook()
```

I think ansible-pull should fail if the user provides a playbook to run and it doesn't exist or there is some other problem reading it.  There may be unintended consequences if we try to be too helpful and try to auto-detect other playbooks.

**Comment by mpdehaan** [comment]
yep, exactly what we want @sfromm

**Comment by mpdehaan** [comment]
I think we can use this as a starting point but I want to tweak this a bit to simplify the code and fix some style things (not all yours, BTW) some before merging.   Stay tuned.

**Comment by copperlight** [comment]
Sounds good, I will keep an eye out.

**Comment by mpdehaan** [comment]
haven't forgot about this, but FYI it will probably be later next week before I get a chance to circle back on this.

**Comment by copperlight** [comment]
Thanks for the update!  I'll continue keeping a lookout for this.  Let me know if you would like testing help or whatnot.

**Comment by sfromm** [comment]
@mpdehaan is there anything i can do to help this move forward?  I'd like to add a --purge option that would remove the git checkout after the ansible-playbook run, but would rather do that after you make your changes/tweaks.

**Comment by mpdehaan** [comment]
I'm quite swamped with organizing and pull requests and fireball mode and
... everything else at the moment.

If you want to test this and run with it, that would be fantastic.

I would like ansible-pull to be less "try-exceptey" and more in line with
the short program it is now.

https://github.com/ansible/ansible/blob/devel/bin/ansible-pull

Where possible, we should move important things into utils (or elsewhere,
if a .pull.py is appropriate, maybe there) and out of the main script body.

Thanks!

**Comment by mpdehaan** [comment]
closing for now, @sfromm said he was working on a patch
